R/install-github.r

```diff
@@ -243,48 +243,67 @@ parse_git_repo <- function(path) {
 }
 
 #' @export
-remote_package_name.github_remote <- function(remote, url = "https://github.com", ...) {
-
+remote_package_name.github_remote <- function(remote, ...) {
   tmp <- tempfile()
-  path <- paste(c(
-      remote$username,
-      remote$repo,
-      "raw",
-      remote$ref,
-      remote$subdir,
-      "DESCRIPTION"), collapse = "/")
-
-  req <- httr::GET(url, path = path, httr::write_disk(path = tmp))
-
-  if (httr::status_code(req) >= 400) {
-    return(NA_character_)
-  }
+  on.exit(unlink(tmp))
+  get_contents_using_github_api(remote,
+                                content_path = "DESCRIPTION",
+                                path_to_save = tmp)
 
   read_dcf(tmp)$Package
 }
 
-#' @export
-remote_sha.github_remote <- function(remote, url = "https://github.com", ...) {
-  # If the remote ref is the same as the sha it is a pinned commit so just
-  # return that.
-  if (!is.null(remote$ref) && !is.null(remote$sha) &&
-    grepl(paste0("^", remote$ref), remote$sha)) {
-    return(remote$sha)
+#' @references https://developer.github.com/v3/repos/contents/#get-contents
+get_contents_using_github_api <-
+  function(remote, content_path, path_to_save) {
+    owner <- remote$username
+    repo <- remote$repo
+    target_path <-
+      file.path("repos", owner, repo, "contents", content_path)
+    response <- github_GET(path = target_path)
+
+    download_file_using_wget(url = response$download_url,
+                             destfile = path_to_save)
   }
 
-  tryCatch({
-    res <- git2r::remote_ls(
-      paste0(url, "/", remote$username, "/", remote$repo, ".git"),
-      ...)
+download_file_using_wget <- function(url, destfile) {
+  utils::download.file(
+    url = url,
+    destfile = destfile,
+    cacheOK = TRUE,
+    method = "wget",
+    quiet = TRUE,
+    extra = "--continue"
+  )
+}
 
-    found <- grep(pattern = paste0("/", remote$ref), x = names(res))
+#' @export
+#' @references https://developer.github.com/v3/repos/commits/#get-the-sha-1-of-a-commit-reference
+remote_sha.github_remote <- function(remote, ...) {
```

Comment:
To sum these up, I'd like to approach with the first suggested method, and it carries an issue rooted from the restriction in the `github_GET`, but I don't want to avoid this problem by not using `github_GET`. Instead, I think it is better to extend `github_GET` together in this time.